tests/python/pants_test/engine/test_engine.py

```diff
@@ -237,6 +238,28 @@ def test_non_existing_root_fails_differently(self):
       str(cm.exception)
     )
 
+  def test_no_cache_exceptions(self):
```

Comment:
It would be good to explain in a code or commit comment why this test demonstrates the new behavior we want out of the engine with this change, since reasoning about pants caching can be tricky. Re-defining the `function_to_call` function mimics a rule that intermittently throws exceptions - the first time we call `product_request` on the scheduler, we hit the `ValueError`-throwing version of `function_to_call` and the whole `product_request` call throws an `ExecutionError`. The second time, we redefine `function_to_call` to not throw an exception - so instead of caching the `ValueError` from the first definition of `function_to_call`, like we would have before this change, the engine re-runs the `f` `@rule`, this time successfully.

Reply:
I've cannibalized your wording here to make a docstring describing why this test works!